core/trino-main/src/main/java/io/trino/execution/scheduler/policy/SequentialExecutionSchedule.java

```diff
@@ -0,0 +1,606 @@
+        activeSources.remove(stage);
+    }
+
+    private void unblockStages()
+    {
+        // find stages that are blocked on full task output buffer
+        Set<PlanFragmentId> blockedFragments = activeSources.stream()
+                .filter(StageExecution::isAnyTaskBlocked)
+                .map(stage -> stage.getFragment().getId())
+                .collect(toImmutableSet());
+        // start immediate downstream stages so that data can be consumed
+        blockedFragments.stream()
+                .flatMap(fragmentId -> fragmentTopology.outgoingEdgesOf(fragmentId).stream())
+                .map(FragmentsEdge::getTarget)
+                .map(stagesByFragmentId::get)
+                .forEach(this::startStage);
+    }
+
+    private void startStage(StageExecution stage)
+    {
+        if (fragmentDependency.outDegreeOf(stage.getFragment().getId()) > 0) {
+            // if there are any dependent stages then reschedule when stage is completed
+            stage.addStateChangeListener(state -> {
+                if (isStageCompleted(stage)) {
+                    notifyReschedule();
+                }
+            });
+        }
+        activeSources.add(stage);
+    }
+
+    private void notifyReschedule()
+    {
+        SettableFuture<Void> rescheduleFuture;
+        synchronized (this) {
+            rescheduleFuture = this.rescheduleFuture;
+            this.rescheduleFuture = SettableFuture.create();
+        }
+        // notify listeners outside of the critical section
+        rescheduleFuture.set(null);
+    }
+
+    private boolean isStageCompleted(StageExecution stage)
+    {
+        State state = stage.getState();
+        return state == SCHEDULED || state == RUNNING || state == FLUSHING || state.isDone();
+    }
+
+    private Set<PlanFragmentId> extractDependencies(Collection<PlanFragment> fragments)
+    {
+        // Build a graph where the plan fragments are vertexes and the edges represent
+        // a before -> after relationship. Destination fragment should be started only
+        // when source fragment is completed. For example, a join hash build has an edge
+        // to the join probe.
+        fragments.forEach(fragment -> {
+            fragmentDependency.addVertex(fragment.getId());
+            fragmentTopology.addVertex(fragment.getId());
+        });
+
+        Set<PlanFragmentId> nonLazyFragments = new HashSet<>();
+        Visitor visitor = new Visitor(fragments, fragmentDependency, fragmentTopology, nonLazyFragments);
+        fragments.forEach(fragment -> visitor.processFragment(fragment.getId()));
+
+        // Make sure there are no strongly connected components as it would mean circular dependency between stages
+        List<Set<PlanFragmentId>> components = new StrongConnectivityInspector<>(fragmentDependency).stronglyConnectedSets();
+        verify(components.stream().allMatch(set -> set.size() == 1), "circular dependency between stages");
```

Comment:
nit: you may just check if nubmer of sets is equal to number of fragments.

Reply:
Also not sure if this check is super beneficial. Aren't we assuming everywhere in the codebase that there are no circular deps between fragments.

Reply:
Something similar is done here `io/trino/execution/scheduler/policy/PhasedExecutionSchedule.java:146`. There broadcast build and probe constitute strongly connected component. Therefore I think the check here makes sense